[RISCV] Remove experimental from Vector Crypto extensions

ebiggers opened the pull request:
The RISC-V vector crypto extensions have been ratified.  This patch updates the Clang and LLVM support for these extensions to be non-experimental, while leaving the C intrinsics as experimental since the C intrinsics are not yet standardized.

This is an updated version of https://github.com/llvm/llvm-project/pull/69000, which hasn't had activity in a few weeks

4vtomat committed:
[RISCV] Remove experimental from Vector Crypto extensions

The RISC-V vector crypto extensions have been ratified.  This patch
updates the Clang and LLVM support for these extensions to be
non-experimental, while leaving the C intrinsics as experimental since
the C intrinsics are not yet standardized.

Co-authored-by: Eric Biggers <ebiggers@google.com>

topperc commented on clang/include/clang/Support/RISCVVIntrinsicUtils.h:
```diff
@@ -485,7 +485,7 @@ class RVVIntrinsic {
 // RVVRequire should be sync'ed with target features, but only
 // required features used in riscv_vector.td.
-enum RVVRequire : uint16_t {
+enum RVVRequire : unsigned int {
```
Use uint32_t since we're counting bits.

ebiggers replied:
Done.  Note, I also changed the type in `RVVIntrinsicRecord::RequiredExtensions` and `SemaRecord::RequiredExtensions`.

topperc commented on clang/lib/Basic/Targets/RISCV.h:
```diff
@@ -28,6 +28,7 @@ class RISCVTargetInfo : public TargetInfo {
 protected:
   std::string ABI, CPU;
   std::unique_ptr<llvm::RISCVISAInfo> ISAInfo;
+  bool HasExperimental = false;
```
Can this be private like `FastUnalignedAccess`?

ebiggers replied:
Done

ebiggers commented on clang/lib/Basic/Targets/RISCV.cpp:
```diff
@@ -370,6 +371,10 @@ bool RISCVTargetInfo::handleTargetFeatures(std::vector<std::string> &Features,
     ISAInfo = std::move(*ParseResult);
   }
 
+  if (std::find(Features.begin(), Features.end(), "+experimental") !=
```
Done using `llvm::is_contained`

ebiggers commented:
> To be safe, I think we should have an experimental tag to guard the intrinsics, while the extension is ratified and do indeed don't need the experimental.

This is already part of the pull request.

I've updated the commit message and pull request description to make it clear that it's still leaving the C intrinsics as experimental.

> What you propose here seems OK to me, but please add something to RISCVUsage indicating that in some cases an extension may be non-experimental but the intrinsics still experimental and -menable-experimental-extensions must be passed to Clang to use them.

I added a new section "Experimental C Intrinsics" to RISCVUsage.rst.

Also in RISCVUsage.rst, I changed the status of the extensions from "Supported" (the previous proposal) to "Assembly Support".  This is consistent with the fact that the file says that "Supported" means the C intrinsics are supported.

> I also think a Clang release note might make sense, such as "It is now possible for intrinsics to be experimental while the extension itself is non-experimental. This is currently the case for the vector crypto extensions (Zvk*, Zvbb, Zvbc) where -menable-experimental-extension must still be passed to Clang in order to enable use of the C intrinsics."

Done, but I combined the information with the note already being added about the vector crypto extensions, since that seems to be more logical.  Let me know what you think; thanks!

ebiggers commented:
The `clang-format` warning is from existing code, not from this pull request.

topperc reviewed (approved):
LGTM

ebiggers commented:
Rebased onto latest upstream because the `clang-format` warning is fixed there.  No changes

topperc commented on llvm/docs/RISCVUsage.rst:
```diff
@@ -143,6 +143,22 @@ on support follow.
      ``Zve64f``       Supported
      ``Zve64d``       Supported
      ``Zvfh``         Supported
+     ``Zvbb``         Assembly Support
```
Please alphabetize this list correctly.

ebiggers replied:
Done

topperc reviewed (approved):
LGTM

asb reviewed (approved):
I was sure I LGTMed this earlier today, but it seems I didn't...

ebiggers commented:
Rebased to resolve conflicts again

topperc commented:
@ebiggers are you table to commit or do you need someone else to do it? I don't recall seeing a commit from you before so I wasn't sure.

ebiggers commented:
No, I don't have commit access.

ebiggers commented:
Updated to add my Co-authored-by, as the original author is `Brandon Wu <brandon.wu@sifive.com>`.

Someone who has permission to do so will need to approve the GitHub Actions workflow, and merge the commit once everything is ready.

ebiggers commented:
Thanks @topperc!
